lib/utils.js

```diff
@@ -339,6 +341,40 @@ function removeUndefined(query, handleUndefined) {
   });
 }
 
+/**
+ * Check the where clause to report prohibited keys
+ * @param {object} where Where object
+ * @param {string[]} [prohibitedKeys] An array of prohibited keys
+ */
+function validateKeys(where, prohibitedKeys) {
+  if (!prohibitedKeys || prohibitedKeys.length === 0) return where;
+  if (typeof where !== 'object' || where === null) {
+    return where;
+  }
+  prohibitedKeys = prohibitedKeys || [];
+  const offendingKeys = [];
+  // WARNING: [rfeng] Use traverse.map() will cause mongodb to produce invalid BSON
+  // as traverse doesn't transform the ObjectId correctly. Instead we have to use
+  // traverse(where).forEach(...)
+  const result = traverse(where).forEach(function(x) {
+    if (prohibitedKeys.indexOf(this.key) !== -1) {
+      offendingKeys.push(this.key);
+    }
+    return x;
+  });
+  if (offendingKeys.length) {
+    const msg = 'Invalid properties are used in query';
+    const err = new Error(msg);
+    err.code = 'PROPERTY_NOT_ALLOWED_IN_QUERY';
+    err.statusCode = 400;
+    err.details = {where: where};
+    debug('Hidden or protected properties %j are used in query: %j',
+      offendingKeys, where, err);
+    throw err;
```

Comment:
By throwing an error we are signaling the client that their request contained a hidden property. As a result, a potential attacker has still a tool to discover hidden properties:
 - Send a query with a single offending key. 
 - If the server returns a `PROPERTY_NOT_ALLOWED_IN_QUERY ` error, we found a hidden model property. Bingo!
 - Otherwise our guess was wrong and we can try another name.

IMO, the only secure solution is to silently ignore hidden/protected fields (from the point of the HTTP/REST client) and log the error for the operator.

I would personally always print the error via `console.error`:
 - at development time, a console error will make it easier for developers to realize they are sending a wrong query
 - in production, a console error log should hopefully trigger investigation and make it more clear the app may be under attack